include/lp_hash_table.h

```diff
@@ -0,0 +1,100 @@
+#ifndef DIL_LP_HASH_TABLE_H
+#define DIL_LP_HASH_TABLE_H
+
+#include <stdlib.h>
+#include <stdint.h>
+#include <assert.h>
+
+template <typename T>
+struct lp_hash_table {
+private:
+    const char **keys;
+    T *values;
+    int buckets;
+
+    unsigned int hash(const char *key) {
+        // TODO: implement a better hashing method
+        return ((uintptr_t)key % buckets);
+    }
+
+public:
+
+    /*
+     * Constructor. Assumes that `elems > 0`
+     */
+    lp_hash_table(int elems) {
+        assert(elems > 0);
+
+        // TODO: maybe find a way to optimize this
+        buckets = elems;
+
+        // TODO: use our own allocator when it is ready.
+        keys = (const char **)malloc(buckets * sizeof(const char *));
+        values = (T *)malloc(buckets * sizeof(T));
```

Comment:
If you want to save `N` elements into a hash table, then using space of `N` is a bad idea. That's because the most important thing in hash table is the good distribution. Good distribution means that when you will _probe_ you will hit very few elements (hence the O(1) amortized). Using space of `N` elements means that essentially this is linear search, i.e. O(n). You have to "waste" some space to achieve good distribution.

That's where the _load factor_ is introduced: A load factor in what percentage is the hash table full. For example, if we're using only 70 out of 100 items, then it's 70%. Usually 70% is the limit for a good load factor. Above that the hash table is very slow (and it needs rebalancing and allocating more elements).

Since we know the number of elements to save, the load factor is constant (still, a good distribution is not guarranteed though). So, we can just take space say `1.3 * elems`;

_Note_: There is a lot more theory (and math) involved with these things than the simplistic explanation I gave but I assumed I'll bore you to death.

Reply:
It's fine, I love big boring explanations